Propositions 1-5 subquieries for Chapter 4

1. Proposition 1 "All Orders placed on the first  day of activity"

In [14]:
USE WideWorldImporters
GO
SELECT * 
FROM Sales.Orders
WHERE OrderDate = (
    SELECT MIN(OrderDate) 
    FROM Sales.Orders
);

Commands completed successfully.

(79 rows affected)

Total execution time: 00:00:00.033

OrderID CustomerID SalespersonPersonID PickedByPersonID ContactPersonID BackorderOrderID OrderDate ExpectedDeliveryDate CustomerPurchaseOrderNumber IsUndersupplyBackordered Comments DeliveryInstructions InternalComments PickingCompletedWhen LastEditedBy LastEditedWhen 1 832 2 NULL 3032 45 2013-01-01 2013-01-02 12126 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 2 803 8 NULL 3003 46 2013-01-01 2013-01-02 15342 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 3 105 7 NULL 1209 47 2013-01-01 2013-01-02 12211 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 4 57 16 3 1113 NULL 2013-01-01 2013-01-02 17129 1 NULL NULL NULL 2013-01-01 11:00:00.0000000 3 2013-01-01 11:00:00.0000000 5 905 3 NULL 3105 48 2013-01-01 2013-01-02 10369 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 6 976 13 3 3176 NULL 2013-01-01 2013-01-02 13383 1 NULL NULL NULL 2013-01-01 11:00:00.0000000 3 2013-01-01 11:00:00.0000000 7 575 8 NULL 2349 49 2013-01-01 2013-01-02 17913 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 8 964 7 NULL 3164 50 2013-01-01 2013-01-02 14518 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 9 77 7 NULL 1153 51 2013-01-01 2013-01-02 17577 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 10 191 20 NULL 1381 52 2013-01-01 2013-01-02 18030 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 11 586 3 NULL 2371 53 2013-01-01 2013-01-02 17639 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 12 529 15 NULL 2257 54 2013-01-01 2013-01-02 10941 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 13 473 13 NULL 2145 55 2013-01-01 2013-01-02 16374 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 14 870 8 NULL 3070 56 2013-01-01 2013-01-02 17521 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 15 991 3 3 3191 NULL 2013-01-01 2013-01-02 11911 1 NULL NULL NULL 2013-01-01 11:00:00.0000000 3 2013-01-01 11:00:00.0000000 16 414 2 NULL 2027 57 2013-01-01 2013-01-02 10690 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 17 910 8 NULL 3110 58 2013-01-01 2013-01-02 10255 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 18 423 2 4 2045 NULL 2013-01-01 2013-01-02 13260 1 NULL NULL NULL 2013-01-02 11:00:00.0000000 4 2013-01-02 11:00:00.0000000 19 949 7 NULL 3149 59 2013-01-01 2013-01-02 16825 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 20 973 14 NULL 3173 60 2013-01-01 2013-01-02 12227 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 21 463 6 4 2125 NULL 2013-01-01 2013-01-02 16070 1 NULL NULL NULL 2013-01-02 11:00:00.0000000 4 2013-01-02 11:00:00.0000000 22 884 13 NULL 3084 61 2013-01-01 2013-01-02 16793 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 23 173 7 NULL 1345 62 2013-01-01 2013-01-02 12467 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 24 890 16 3 3090 NULL 2013-01-01 2013-01-02 13700 1 NULL NULL NULL 2013-01-01 11:00:00.0000000 3 2013-01-01 11:00:00.0000000 25 189 15 NULL 1377 63 2013-01-01 2013-01-02 11503 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 26 873 15 3 3073 NULL 2013-01-01 2013-01-02 13916 1 NULL NULL NULL 2013-01-01 11:00:00.0000000 3 2013-01-01 11:00:00.0000000 27 497 6 NULL 2193 64 2013-01-01 2013-01-02 17783 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 28 484 20 NULL 2167 65 2013-01-01 2013-01-02 13042 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 29 89 8 NULL 1177 66 2013-01-01 2013-01-02 10411 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:00.0000000 30 148 6 NULL 1295 67 2013-01-01 2013-01-02 15084 1 NULL NULL NULL 2013-01-01 12:00:00.0000000 7 2013-01-01 12:00:

<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">Proposition 2&nbsp;</span>   <span style="color: rgb(0, 128, 0); font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace; font-size: 12px; white-space: pre;">"</span><span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">Query of customers that returns orders after jan 12,2013</span><span style="color: rgb(0, 128, 0); font-family: Menlo, Monaco, &quot;Courier New&quot;, monospace; font-size: 12px; white-space: pre;">"</span>

In [17]:
--code 2 
USE WideWorldImporters
GO
SELECT DISTINCT CustomerID, CustomerName
FROM Sales.Customers
WHERE CustomerID IN (
    SELECT CustomerID 
    FROM Sales.Orders
    WHERE OrderDate > '2013-01-12'
);

Commands completed successfully.

(663 rows affected)

Total execution time: 00:00:00.039

CustomerID CustomerName 1 Tailspin Toys (Head Office) 2 Tailspin Toys (Sylvanite, MT) 3 Tailspin Toys (Peeples Valley, AZ) 4 Tailspin Toys (Medicine Lodge, KS) 5 Tailspin Toys (Gasport, NY) 6 Tailspin Toys (Jessie, ND) 7 Tailspin Toys (Frankewing, TN) 8 Tailspin Toys (Bow Mar, CO) 9 Tailspin Toys (Netcong, NJ) 10 Tailspin Toys (Wimbledon, ND) 11 Tailspin Toys (Devault, PA) 12 Tailspin Toys (Biscay, MN) 13 Tailspin Toys (Stonefort, IL) 14 Tailspin Toys (Long Meadow, MD) 15 Tailspin Toys (Batson, TX) 16 Tailspin Toys (Coney Island, MO) 17 Tailspin Toys (East Fultonham, OH) 18 Tailspin Toys (Goffstown, NH) 19 Tailspin Toys (Lemeta, AK) 20 Tailspin Toys (College Place, WA) 21 Tailspin Toys (Tresckow, PA) 22 Tailspin Toys (Ward Ridge, FL) 23 Tailspin Toys (Ikatan, AK) 24 Tailspin Toys (Dundarrach, NC) 25 Tailspin Toys (Avenal, CA) 26 Tailspin Toys (Hedrick, IA) 27 Tailspin Toys (Bowlus, MN) 28 Tailspin Toys (North Ridge, NY) 29 Tailspin Toys (Eulaton, AL) 30 Tailspin Toys (Koontzville, WA) 31 Tailspin Toys (Lake Stevens, WA) 32 Tailspin Toys (Maypearl, TX) 33 Tailspin Toys (Boyden Arbor, SC) 34 Tailspin Toys (Dallardsville, TX) 35 Tailspin Toys (Slanesville, WV) 36 Tailspin Toys (Ekron, KY) 37 Tailspin Toys (Kerby, OR) 38 Tailspin Toys (Airport Drive, MO) 39 Tailspin Toys (Diablock, KY) 40 Tailspin Toys (Impact, TX) 41 Tailspin Toys (Olivette, MO) 42 Tailspin Toys (Arietta, NY) 43 Tailspin Toys (Upper Preston, WA) 44 Tailspin Toys (Amanda Park, WA) 45 Tailspin Toys (Severna Park, MD) 46 Tailspin Toys (Jemison, AL) 47 Tailspin Toys (Lake Hughes, CA) 48 Tailspin Toys (Trentwood, WA) 49 Tailspin Toys (Muir, MI) 50 Tailspin Toys (Hambleton, WV) 51 Tailspin Toys (Imlaystown, NJ) 52 Tailspin Toys (Idria, CA) 53 Tailspin Toys (Nanafalia, AL) 54 Tailspin Toys (Railroad, PA) 55 Tailspin Toys (Gum Branch, GA) 56 Tailspin Toys (Royal City, WA) 57 Tailspin Toys (Lakemore, OH) 58 Tailspin Toys (Mappsburg, VA) 59 Tailspin Toys (Edgartown, MA) 60 Tailspin Toys (East Dailey, WV) 61 Tailspin Toys (Fairfield Glade, TN) 62 Tailspin Toys (Great Neck Estates, NY) 63 Tailspin Toys (Stallion Springs, CA) 64 Tailspin Toys (Hodgdon, ME) 65 Tailspin Toys (North Cowden, TX) 66 Tailspin Toys (Madaket, MA) 67 Tailspin Toys (Sentinel Butte, ND) 68 Tailspin Toys (Gardere, LA) 69 Tailspin Toys (Lytle Creek, CA) 70 Tailspin Toys (New Baden, IL) 71 Tailspin Toys (Good Hart, MI) 72 Tailspin Toys (Cortaro, AZ) 73 Tailspin Toys (Cokato, MN) 74 Tailspin Toys (Indios, PR) 75 Tailspin Toys (Windsor Locks, CT) 76 Tailspin Toys (Yewed, OK) 77 Tailspin Toys (Hahira, GA) 78 Tailspin Toys (Sekiu, WA) 79 Tailspin Toys (Page City, KS) 80 Tailspin Toys (Valdese, NC) 81 Tailspin Toys (Big Moose, NY) 82 Tailspin Toys (La Cueva, NM) 83 Tailspin Toys (Absecon, NJ) 84 Tailspin Toys (Aceitunas, PR) 85 Tailspin Toys (Andrix, CO) 86 Tailspin Toys (New Lexington, OH) 87 Tailspin Toys (Sauquoit, NY) 88 Tailspin Toys (Dracut, MA) 89 Tailspin Toys (Victory Gardens, NJ) 90 Tailspin Toys (Tolna, ND) 91 Tailspin Toys (Alstead, NH) 92 Tailspin Toys (Sans Souci, SC) 93 Tailspin Toys (Clewiston, FL) 94 Tailspin Toys (Cheyenne Wells, CO) 95 Tailspin Toys (Lavon, TX) 96 Tailspin Toys (South La Paloma, TX) 97 Tailspin Toys (Manchester Center, VT) 98 Tailspin Toys (El Centro, CA) 99 Tailspin Toys (Guin, AL) 100 Tailspin Toys (Tunnelhill, PA) 101 Tailspin Toys (Cundiyo, NM) 102 Tailspin Toys (Fieldbrook, CA) 103 Tailspin Toys (Kalvesta, KS) 104 Tailspin Toys (Wallagrass, ME) 105 Tailspin Toys (Tomnolen, MS) 106 Tailspin Toys (Tumacacori, AZ) 107 Tailspin Toys (Glen Avon, CA) 108 Tailspin Toys (Bernie, MO) 109 Tailspin Toys (South Laguna, CA) 110 Tailspin Toys (North Crows Nest, IN) 111 Tailspin Toys (Oriole Beach, FL) 112 Tailspin Toys (Sallyards, KS) 113 Tailspin Toys (Dahlia, NM) 114 Tailspin Toys (Cherry Grove Beach, SC) 115 Tailspin Toys (Bethania, NC) 116 Tailspin Toys (Rafael Capó, PR) 117 Tailspin Toys (Scofield, UT) 118 Tailspin Toys (Vassalboro, ME) 119 Tailspin Toys (Statenville, GA) 120 Tai

```
Propostion 3 "Query that returns employees who proccessed orders on or after july 2013"

```

In [21]:
--code 3
USE WideWorldImporters
GO
SELECT DISTINCT PersonID AS EmployeeID, FullName AS EmployeeName
FROM Application.People
WHERE PersonID IN (
    SELECT SalespersonPersonID
    FROM Sales.Orders
    WHERE OrderDate >= '2013-07-01'
);

Commands completed successfully.

(10 rows affected)

Total execution time: 00:00:00.028

EmployeeID EmployeeName 15 Taj Shand 3 Hudson Onslow 6 Sophia Hinton 7 Amy Trefl 13 Hudson Hollinworth 16 Archer Lamble 2 Kayla Woodcock 20 Jack Potter 14 Lily Code 8 Anthony Grosse

Proposition 4 "Query that returns all orders placed by the customers who placed the highest number of order"

In [22]:
--code 4
USE WideWorldImporters
GO
SELECT CustomerID, OrderID
FROM Sales.Orders
WHERE CustomerID IN (
    SELECT CustomerID
    FROM Sales.Orders
    GROUP BY CustomerID
    HAVING COUNT(OrderID) = (
        SELECT MAX(MaxOrderCount)
        FROM (
            SELECT CustomerID, COUNT(OrderID) AS MaxOrderCount
            FROM Sales.Orders
            GROUP BY CustomerID
        ) AS OrderCounts
    )
);

Commands completed successfully.

(150 rows affected)

Total execution time: 00:00:00.023

CustomerID OrderID 90 1455 90 1890 90 2648 90 2865 90 2878 90 3248 90 3455 90 3521 90 3881 90 3926 90 4626 90 5693 90 5757 90 6638 90 7139 90 7854 90 10807 90 10965 90 12044 90 12470 90 12849 90 12963 90 13249 90 13864 90 15460 90 15470 90 15493 90 15515 90 16678 90 17032 90 17059 90 17546 90 18096 90 18358 90 18381 90 18416 90 18464 90 18635 90 19013 90 19063 90 19430 90 20023 90 20735 90 21481 90 24292 90 24407 90 24439 90 25524 90 26509 90 27266 90 27292 90 28044 90 28103 90 29989 90 30303 90 31656 90 32207 90 32217 90 32263 90 32302 90 32370 90 32986 90 33348 90 33439 90 33685 90 35735 90 36079 90 36099 90 36188 90 36225 90 36249 90 36270 90 36346 90 36424 90 37449 90 38828 90 39745 90 40500 90 40576 90 40578 90 41300 90 41308 90 41596 90 41939 90 42337 90 43231 90 43604 90 43624 90 44461 90 45458 90 45927 90 46146 90 46297 90 46601 90 46672 90 48011 90 49100 90 49139 90 49603 90 49780 90 50332 90 51039 90 51346 90 51353 90 52841 90 53287 90 53482 90 53531 90 53974 90 54353 90 54460 90 54518 90 54595 90 54896 90 54908 90 55759 90 56153 90 56570 90 56774 90 57518 90 57552 90 58292 90 59179 90 60224 90 60251 90 60543 90 60555 90 60587 90 61421 90 61698 90 62833 90 63521 90 63679 90 63794 90 65604 90 65669 90 66593 90 66613 90 68393 90 70091 90 70215 90 70729 90 70928 90 71287 90 71450 90 71940 90 71999 90 72412 90 72445 90 73565

Proposition 5 "Query that Returns orders placed by employees starting wthe letter A" reference page 152 for help

In [5]:
--code 5
USE WideWorldImporters
GO

SELECT 
    od.OrderID,
    od.OrderDate,
    p.FullName AS [Employee Name] 
FROM Sales.Orders od
JOIN Application.People p 
ON od.SalespersonPersonID = p.PersonID
WHERE p.FullName LIKE 'A%'; 

Commands completed successfully.

(22065 rows affected)

Displaying Top 5000 rows.

Total execution time: 00:00:00.045

OrderID OrderDate Employee Name 2 2013-01-01 Anthony Grosse 3 2013-01-01 Amy Trefl 4 2013-01-01 Archer Lamble 7 2013-01-01 Anthony Grosse 8 2013-01-01 Amy Trefl 9 2013-01-01 Amy Trefl 14 2013-01-01 Anthony Grosse 17 2013-01-01 Anthony Grosse 19 2013-01-01 Amy Trefl 23 2013-01-01 Amy Trefl 24 2013-01-01 Archer Lamble 29 2013-01-01 Anthony Grosse 35 2013-01-01 Anthony Grosse 38 2013-01-01 Archer Lamble 42 2013-01-01 Anthony Grosse 46 2013-01-01 Anthony Grosse 47 2013-01-01 Amy Trefl 49 2013-01-01 Anthony Grosse 50 2013-01-01 Amy Trefl 51 2013-01-01 Amy Trefl 56 2013-01-01 Anthony Grosse 58 2013-01-01 Anthony Grosse 59 2013-01-01 Amy Trefl 62 2013-01-01 Amy Trefl 66 2013-01-01 Anthony Grosse 71 2013-01-01 Anthony Grosse 73 2013-01-01 Archer Lamble 77 2013-01-01 Anthony Grosse 82 2013-01-02 Amy Trefl 85 2013-01-02 Anthony Grosse 87 2013-01-02 Archer Lamble 90 2013-01-02 Anthony Grosse 91 2013-01-02 Amy Trefl 92 2013-01-02 Amy Trefl 95 2013-01-02 Archer Lamble 96 2013-01-02 Anthony Grosse 98 2013-01-02 Anthony Grosse 108 2013-01-02 Archer Lamble 114 2013-01-02 Amy Trefl 117 2013-01-02 Archer Lamble 123 2013-01-02 Archer Lamble 127 2013-01-02 Anthony Grosse 128 2013-01-02 Amy Trefl 133 2013-01-02 Archer Lamble 134 2013-01-02 Anthony Grosse 136 2013-01-02 Archer Lamble 137 2013-01-02 Anthony Grosse 146 2013-01-02 Amy Trefl 150 2013-01-02 Archer Lamble 154 2013-01-03 Archer Lamble 162 2013-01-03 Anthony Grosse 164 2013-01-03 Anthony Grosse 173 2013-01-03 Anthony Grosse 174 2013-01-03 Archer Lamble 175 2013-01-03 Anthony Grosse 176 2013-01-03 Anthony Grosse 178 2013-01-03 Anthony Grosse 179 2013-01-03 Amy Trefl 180 2013-01-03 Amy Trefl 183 2013-01-03 Anthony Grosse 193 2013-01-03 Amy Trefl 194 2013-01-03 Anthony Grosse 198 2013-01-03 Archer Lamble 199 2013-01-03 Amy Trefl 200 2013-01-03 Amy Trefl 203 2013-01-03 Amy Trefl 207 2013-01-03 Archer Lamble 210 2013-01-03 Amy Trefl 214 2013-01-03 Anthony Grosse 216 2013-01-03 Anthony Grosse 221 2013-01-03 Anthony Grosse 225 2013-01-03 Anthony Grosse 230 2013-01-03 Amy Trefl 231 2013-01-03 Amy Trefl 232 2013-01-03 Amy Trefl 235 2013-01-03 Amy Trefl 238 2013-01-04 Archer Lamble 245 2013-01-04 Archer Lamble 246 2013-01-04 Anthony Grosse 253 2013-01-04 Amy Trefl 254 2013-01-04 Archer Lamble 255 2013-01-04 Archer Lamble 256 2013-01-04 Amy Trefl 257 2013-01-04 Anthony Grosse 264 2013-01-04 Amy Trefl 266 2013-01-04 Amy Trefl 268 2013-01-04 Anthony Grosse 270 2013-01-03 Amy Trefl 271 2013-01-04 Archer Lamble 274 2013-01-04 Anthony Grosse 277 2013-01-04 Anthony Grosse 279 2013-01-05 Anthony Grosse 280 2013-01-05 Archer Lamble 284 2013-01-05 Amy Trefl 286 2013-01-05 Archer Lamble 294 2013-01-05 Archer Lamble 295 2013-01-05 Amy Trefl 297 2013-01-05 Archer Lamble 299 2013-01-05 Archer Lamble 300 2013-01-05 Amy Trefl 303 2013-01-05 Amy Trefl 306 2013-01-05 Anthony Grosse 319 2013-01-05 Anthony Grosse 320 2013-01-05 Amy Trefl 322 2013-01-05 Archer Lamble 329 2013-01-07 Amy Trefl 332 2013-01-07 Anthony Grosse 335 2013-01-07 Amy Trefl 339 2013-01-07 Archer Lamble 341 2013-01-07 Amy Trefl 343 2013-01-07 Archer Lamble 344 2013-01-07 Amy Trefl 352 2013-01-07 Archer Lamble 355 2013-01-07 Anthony Grosse 356 2013-01-07 Amy Trefl 362 2013-01-07 Anthony Grosse 365 2013-01-07 Amy Trefl 366 2013-01-07 Anthony Grosse 369 2013-01-07 Amy Trefl 372 2013-01-07 Archer Lamble 376 2013-01-07 Archer Lamble 378 2013-01-07 Archer Lamble 382 2013-01-07 Archer Lamble 388 2013-01-07 Archer Lamble 389 2013-01-07 Anthony Grosse 393 2013-01-07 Archer Lamble 397 2013-01-07 Amy Trefl 400 2013-01-07 Anthony Grosse 403 2013-01-07 Anthony Grosse 407 2013-01-07 Anthony Grosse 412 2013-01-07 Archer Lamble 414 2013-01-07 Anthony Grosse 415 2013-01-07 Archer Lamble 416 2013-01-07 Anthony Grosse 417 2013-01-07 Archer Lamble 419 2013-01-07 Anthony Grosse 424 2013-01-07 Archer Lamble 426 2013-01-07 Anthony Grosse 427 2013-01-08 Anthony Grosse 428 2013-01-08 Anthony Grosse 429 2013-01-08 Archer Lamble 440 2013-01-08 Archer Lamble 458 2013-01-08 Arc

Propositions 6-10 | Table Expressions | Chap 5

Propistion 6 "sales per year"

In [18]:
--code 6 -

USE WideWorldImporters
GO 
WITH YearlySales AS   (
    SELECT 
        YEAR(od.OrderDate) AS SalesYear, --column alias
        COUNT(od.OrderID) AS TotalSales,   --CA
        SUM(ol.Quantity) AS TotalQuantity  --CA
    FROM Sales.Orders od
    JOIN Sales.OrderLines ol ON od.OrderID = ol.OrderID
    GROUP BY YEAR(od.OrderDate)
)
SELECT * FROM YearlySales
ORDER BY SalesYear DESC;

Commands completed successfully.

(4 rows affected)

Total execution time: 00:00:00.062

SalesYear TotalSales TotalQuantity 2016 29902 1291028 2015 73003 2873250 2014 66852 2672973 2013 61655 2473653

Proposition 7 "returns the maximum(recent) value in the orderdate column for each customer " order

In [23]:

USE WideWorldImporters
GO 
SELECT 
    CustomerID, 
    MAX(OrderDate) AS LastOrderDate
FROM Sales.Orders
GROUP BY CustomerID;

Commands completed successfully.

(663 rows affected)

Total execution time: 00:00:00.024

CustomerID LastOrderDate 925 2016-05-31 593 2016-05-28 902 2016-05-23 570 2016-05-31 23 2016-05-27 547 2016-05-13 46 2016-05-27 879 2016-05-06 1042 2016-05-09 524 2016-05-25 856 2016-05-31 401 2016-05-26 192 2016-05-23 69 2016-05-24 92 2016-05-13 430 2016-05-20 501 2016-05-28 115 2016-05-20 138 2016-05-19 802 2016-05-31 971 2016-05-24 1048 2016-05-27 407 2016-05-24 948 2016-05-25 825 2016-05-16 161 2016-05-26 919 2016-05-31 1002 2016-05-27 908 2016-05-20 576 2016-05-30 29 2016-05-31 564 2016-05-20 965 2016-05-27 418 2016-05-27 169 2016-05-31 175 2016-05-24 424 2016-05-28 1011 2016-05-20 816 2016-04-29 470 2016-05-26 862 2016-05-30 75 2016-05-06 842 2016-05-20 865 2016-05-19 201 2016-05-12 464 2016-05-20 132 2016-05-26 1054 2016-05-30 487 2016-05-14 438 2016-05-30 181 2016-04-29 507 2016-05-20 9 2016-05-24 15 2016-04-13 109 2016-05-31 158 2016-05-27 822 2016-05-31 458 2016-05-25 556 2016-05-20 481 2016-05-27 530 2016-05-23 89 2016-05-30 587 2016-05-30 848 2016-05-30 836 2016-05-26 195 2016-05-28 550 2016-05-24 991 2016-05-26 979 2016-04-29 942 2016-05-13 1028 2016-05-31 3 2016-05-30 152 2016-05-09 146 2016-05-26 52 2016-05-17 544 2016-05-26 885 2016-05-27 985 2016-05-27 493 2016-05-27 444 2016-05-30 1034 2016-05-26 427 2016-05-27 95 2016-05-17 189 2016-05-19 521 2016-05-20 72 2016-05-27 404 2016-05-31 473 2016-05-30 805 2016-05-14 450 2016-05-26 118 2016-05-14 922 2016-05-11 999 2016-05-12 928 2016-05-11 951 2016-05-24 945 2016-05-30 876 2016-05-23 1045 2016-05-17 899 2016-05-27 1022 2016-05-13 172 2016-05-26 421 2016-04-27 1014 2016-05-31 66 2016-04-11 78 2016-05-16 527 2016-05-19 410 2016-05-24 859 2016-05-18 968 2016-05-19 1051 2016-05-28 905 2016-03-31 32 2016-05-24 573 2016-04-28 916 2016-05-13 513 2016-05-09 1005 2016-05-13 467 2016-04-14 126 2016-05-30 959 2016-05-30 962 2016-05-16 567 2016-05-30 26 2016-05-04 813 2016-05-31 178 2016-05-03 510 2016-05-30 12 2016-05-27 1008 2016-05-09 533 2016-05-17 35 2016-05-31 1031 2016-05-27 129 2016-05-23 461 2016-05-24 106 2016-05-25 155 2016-05-12 819 2016-05-04 868 2016-05-25 86 2016-05-27 845 2016-05-17 584 2016-05-31 484 2016-04-18 135 2016-05-16 553 2016-05-05 112 2016-05-23 63 2016-05-14 6 2016-05-31 504 2016-05-26 931 2016-05-26 896 2016-05-25 1025 2016-05-31 43 2016-05-13 55 2016-05-20 882 2016-05-26 833 2016-05-04 98 2016-05-20 590 2016-05-18 839 2016-05-05 596 2016-05-31 939 2016-05-27 447 2016-05-06 441 2016-05-19 982 2016-05-20 988 2016-05-26 490 2016-05-31 49 2016-05-20 888 2016-05-24 149 2016-05-24 198 2016-05-31 854 2016-05-19 190 2016-04-25 522 2016-05-11 545 2016-05-27 877 2016-05-24 144 2016-05-27 476 2016-05-12 808 2016-05-16 831 2016-05-24 167 2016-05-28 499 2016-05-14 946 2016-04-19 1017 2016-04-30 1023 2016-05-18 1000 2016-05-28 453 2016-05-11 994 2016-05-26 405 2016-05-20 900 2016-05-31 923 2016-05-13 1046 2016-05-27 860 2016-05-31 184 2016-05-19 516 2016-05-26 528 2016-05-13 113 2016-05-19 422 2016-05-19 871 2016-05-17 67 2016-05-23 562 2016-05-27 954 2016-05-19 917 2016-05-17 27 2016-05-25 21 2016-05-27 127 2016-05-05 121 2016-05-24 963 2016-05-30 1006 2016-05-11 814 2016-05-19 568 2016-05-25 1057 2016-05-27 173 2016-05-19 960 2016-05-02 468 2016-05-31 1009 2016-05-06 107 2016-05-26 914 2016-05-07 58 2016-05-26 820 2016-05-18 485 2016-05-19 156 2016-05-27 891 2016-05-31 130 2016-04-13 413 2016-05-28 81 2016-05-27 536 2016-05-27 462 2016-05-24 436 2016-05-26 64 2016-05-31 38 2016-05-18 579 2016-05-28 87 2016-05-31 505 2016-05-25 599 2016-04-28 101 2016-05-10 1026 2016-05-20 883 2016-05-23 977 2016-05-27 585 2016-05-13 44 2016-05-30 7 2016-05-27 1040 2016-05-30 442 2016-05-27 193 2016-05-16 934 2016-05-17 940 2016-05-30 834 2016-05-30 840 2016-05-31 983 2016-05-31 150 2016-05-31 542 2016-05-14 591 2016-05-25 50 2016-05-24 1 2016-05-27 24 2016-05-24 1020 2016-05-09 926 2016-05-16 47 2016-05-30 1043 2016-05-23 141 2016-05-19 903 2016-05-23 571 2016-05-03 70 2016-05-18 402 2016-05-21 425 2016-05-26 548 2016-05-30 93 2016-04-27 880 2016-05-13 525 2016-05-12 851 20

Proposition 8. "Create a view that returns the total quantity for each employee and year"

In [27]:
--code 8 ---not workung 
USE WideWorldImporters

DROP VIEW IF EXISTS EmployeeYearlySales;
GO

CREATE VIEW EmployeeYearlySales AS
SELECT 
    o.SalespersonpersonID AS [Employee ID],
    YEAR(o.OrderDate) AS [Year],
    SUM(ol.Quantity) AS [Total Quantity Sold]
FROM Sales.Orders o
JOIN Sales.OrderLines ol ON o.OrderID = ol.OrderID
GROUP BY o.SalespersonpersonID, YEAR(o.OrderDate);

: Msg 111, Level 15, State 1, Line 6
'CREATE VIEW' must be the first statement in a query batch.

Total execution time: 00:00:00.002

Proposition 9 "<span style="font-family: -apple-system, BlinkMacSystemFont, sans-serif; color: var(--vscode-foreground);">Create a view that returns the total quantity for each customer and year</span>"

In [35]:
--code 9

USE WideWorldImporters
DROP VIEW IF EXISTS CustomerYearlySales;
GO

CREATE VIEW CustomerYearlySales AS
WITH YearlySales AS (
    SELECT 
        od.CustomerID,
        YEAR(od.OrderDate) AS SalesYear,
        SUM(ol.Quantity) AS TotalQuantity
    FROM Sales.Orders od
    JOIN Sales.OrderLines ol ON od.OrderID = ol.OrderID
    GROUP BY od.CustomerID, YEAR(od.OrderDate)
)
SELECT * FROM YearlySales;

Commands completed successfully.

Commands completed successfully.

Total execution time: 00:00:00.010

Proposition 10 "Return each customer's most recent order details and expecteddeliverydates" **\--cross apply pg 198**

In [26]:
--code 10
USE WideWorldImporters
GO  
SELECT 
    c.CustomerID, 
    c.CustomerName, 
    o.OrderID, 
    o.OrderDate, 
    o.ExpectedDeliveryDate
FROM Sales.Customers c
CROSS APPLY (
    SELECT TOP 1 
        o.OrderID, 
        o.OrderDate, 
        o.ExpectedDeliveryDate
    FROM Sales.Orders o
    WHERE o.CustomerID = c.CustomerID
    ORDER BY o.OrderDate DESC
) o;

Commands completed successfully.

(663 rows affected)

Total execution time: 00:00:00.148

CustomerID CustomerName OrderID OrderDate ExpectedDeliveryDate 1 Tailspin Toys (Head Office) 73290 2016-05-27 2016-05-30 2 Tailspin Toys (Sylvanite, MT) 72475 2016-05-14 2016-05-16 3 Tailspin Toys (Peeples Valley, AZ) 73465 2016-05-30 2016-05-31 4 Tailspin Toys (Medicine Lodge, KS) 71366 2016-04-28 2016-04-29 5 Tailspin Toys (Gasport, NY) 73407 2016-05-28 2016-05-30 6 Tailspin Toys (Jessie, ND) 73547 2016-05-31 2016-06-01 7 Tailspin Toys (Frankewing, TN) 73365 2016-05-27 2016-05-30 8 Tailspin Toys (Bow Mar, CO) 72847 2016-05-20 2016-05-23 9 Tailspin Toys (Netcong, NJ) 73059 2016-05-24 2016-05-25 10 Tailspin Toys (Wimbledon, ND) 73045 2016-05-24 2016-05-25 11 Tailspin Toys (Devault, PA) 73573 2016-05-31 2016-06-01 12 Tailspin Toys (Biscay, MN) 73308 2016-05-27 2016-05-30 13 Tailspin Toys (Stonefort, IL) 73213 2016-05-26 2016-05-27 14 Tailspin Toys (Long Meadow, MD) 73486 2016-05-30 2016-05-31 15 Tailspin Toys (Batson, TX) 70298 2016-04-13 2016-04-14 16 Tailspin Toys (Coney Island, MO) 73287 2016-05-27 2016-05-30 17 Tailspin Toys (East Fultonham, OH) 72661 2016-05-18 2016-05-19 18 Tailspin Toys (Goffstown, NH) 73073 2016-05-24 2016-05-25 19 Tailspin Toys (Lemeta, AK) 72054 2016-05-09 2016-05-10 20 Tailspin Toys (College Place, WA) 72914 2016-05-20 2016-05-23 21 Tailspin Toys (Tresckow, PA) 73354 2016-05-27 2016-05-30 22 Tailspin Toys (Ward Ridge, FL) 73163 2016-05-25 2016-05-26 23 Tailspin Toys (Ikatan, AK) 73370 2016-05-27 2016-05-30 24 Tailspin Toys (Dundarrach, NC) 73074 2016-05-24 2016-05-25 25 Tailspin Toys (Avenal, CA) 72079 2016-05-09 2016-05-10 26 Tailspin Toys (Hedrick, IA) 71717 2016-05-04 2016-05-05 27 Tailspin Toys (Bowlus, MN) 73129 2016-05-25 2016-05-26 28 Tailspin Toys (North Ridge, NY) 73519 2016-05-31 2016-06-01 29 Tailspin Toys (Eulaton, AL) 73557 2016-05-31 2016-06-01 30 Tailspin Toys (Koontzville, WA) 71472 2016-04-29 2016-05-02 31 Tailspin Toys (Lake Stevens, WA) 73269 2016-05-26 2016-05-27 32 Tailspin Toys (Maypearl, TX) 73084 2016-05-24 2016-05-25 33 Tailspin Toys (Boyden Arbor, SC) 73341 2016-05-27 2016-05-30 34 Tailspin Toys (Dallardsville, TX) 72791 2016-05-19 2016-05-20 35 Tailspin Toys (Slanesville, WV) 73507 2016-05-31 2016-06-01 36 Tailspin Toys (Ekron, KY) 72863 2016-05-20 2016-05-23 37 Tailspin Toys (Kerby, OR) 72769 2016-05-19 2016-05-20 38 Tailspin Toys (Airport Drive, MO) 72675 2016-05-18 2016-05-19 39 Tailspin Toys (Diablock, KY) 72909 2016-05-20 2016-05-23 40 Tailspin Toys (Impact, TX) 73207 2016-05-26 2016-05-27 41 Tailspin Toys (Olivette, MO) 72253 2016-05-11 2016-05-12 42 Tailspin Toys (Arietta, NY) 73434 2016-05-30 2016-05-31 43 Tailspin Toys (Upper Preston, WA) 72378 2016-05-13 2016-05-16 44 Tailspin Toys (Amanda Park, WA) 73429 2016-05-30 2016-05-31 45 Tailspin Toys (Severna Park, MD) 73151 2016-05-25 2016-05-26 46 Tailspin Toys (Jemison, AL) 73363 2016-05-27 2016-05-30 47 Tailspin Toys (Lake Hughes, CA) 73469 2016-05-30 2016-05-31 48 Tailspin Toys (Trentwood, WA) 73125 2016-05-25 2016-05-26 49 Tailspin Toys (Muir, MI) 72929 2016-05-20 2016-05-23 50 Tailspin Toys (Hambleton, WV) 73056 2016-05-24 2016-05-25 51 Tailspin Toys (Imlaystown, NJ) 73447 2016-05-30 2016-05-31 52 Tailspin Toys (Idria, CA) 72607 2016-05-17 2016-05-18 53 Tailspin Toys (Nanafalia, AL) 73132 2016-05-25 2016-05-26 54 Tailspin Toys (Railroad, PA) 73454 2016-05-30 2016-05-31 55 Tailspin Toys (Gum Branch, GA) 72925 2016-05-20 2016-05-23 56 Tailspin Toys (Royal City, WA) 72745 2016-05-19 2016-05-20 57 Tailspin Toys (Lakemore, OH) 73185 2016-05-26 2016-05-27 58 Tailspin Toys (Mappsburg, VA) 73226 2016-05-26 2016-05-27 59 Tailspin Toys (Edgartown, MA) 73358 2016-05-27 2016-05-30 60 Tailspin Toys (East Dailey, WV) 72208 2016-05-11 2016-05-12 61 Tailspin Toys (Fairfield Glade, TN) 73002 2016-05-23 2016-05-24 62 Tailspin Toys (Great Neck Estates, NY) 73470 2016-05-30 2016-05-31 63 Tailspin Toys (Stallion Springs, CA) 72498 2016-05-14 2016-05-16 64 Tailspin Toys (Hodgdon, ME) 73561 2016-05-31 2016-06-01 65 Tailspin Toys (North 